src/java.base/share/classes/java/math/MutableBigInteger.java

```diff
@@ -1882,6 +1906,204 @@ MutableBigInteger sqrt() {
         }
     }
 
+    /**
+     * Calculate the integer square root {@code floor(sqrt(this))} and the remainder
+     * if needed, where {@code sqrt(.)} denotes the mathematical square root.
+     * The contents of {@code this} are <b>not</b> changed.
+     * The value of {@code this} is assumed to be non-negative.
+     * @throws ArithmeticException if the value returned by {@code bitLength()}
+     * overflows the range of {@code int}.
+     * @return the integer square root of {@code this} and the remainder if needed
+     */
+    MutableBigInteger[] sqrtRem(boolean needRemainder) {
+        // Special cases.
+        if (this.isZero()) {
+            return new MutableBigInteger[] {
+                    new MutableBigInteger(), needRemainder ? new MutableBigInteger() : null
+            };
+        } else if (this.intLen <= 2) {
+            final long x = this.toLong(); // unsigned
+
+            if (this.intLen == 1) {
+                if (x < 4) // result is unity
+                    return new MutableBigInteger[] {
+                            ONE, needRemainder ? new MutableBigInteger((int) (x - 1)) : null
+                    };
+
+                long s = (long) Math.sqrt(x); // Math.sqrt is exact
+                return new MutableBigInteger[] {
+                        new MutableBigInteger((int) s),
+                        needRemainder ? new MutableBigInteger((int) (x - s * s)) : null
+                };
+            }
+
+            // Initial estimate is the square root of the unsigned long value.
+            long s = (long) Math.sqrt(x > 0 ? x : x + 0x1p64);
+            if (s > LONG_MASK // avoid overflow of s * s
+                    || Long.compareUnsigned(x, s * s) < 0) {
+                s--;
+            }
+
+            return new MutableBigInteger[] {
+                    new MutableBigInteger((int) s),
+                    needRemainder ? new MutableBigInteger(x - s * s) : null
+            };
+        }
+
+        long bitLength = this.bitLength();
+        if (bitLength != (int) bitLength)
+            throw new ArithmeticException("bitLength() integer overflow");
+
+        // Normalize
+        MutableBigInteger x = this;
+        int shift = Integer.numberOfLeadingZeros(x.value[x.offset]) & ~1; // shift must be even
+        if (shift != 0) {
+            x = new MutableBigInteger(x);
+            x.primitiveLeftShift(shift);
+        }
+
+        MutableBigInteger[] sqrtRem = x.sqrtRemZimmermann(x.intLen, needRemainder);
+
+        // Unnormalize
```

Comment:
This unnormalization code, as well as the one in the next method, requires a full explanation, perhaps in abstract, mathematical terms, to help understanding.

Reply:
The full explanation for the unnormalization is in the second paper, "A proof of GMP square root", par. 3.2 at page 11.

Reply:
Well, I have to compare that section, which is clear to me, with the code again.

Reply:
@rgiulietti I noticed that, after this unnormalization, while the square root can't have leading zeros by construction, the remainder can have them. Although there's not a general contract of the class MutableBigInteger that imposes to its methods that their results must be always normalized (in the sense of `MBI.isNormalized()`), this might be an issue if `MBI.sqrtRem()` were used to do intermediate computations with MutableBigIntegers.